packages/backend/src/open_payments/payment/outgoing/service.ts

```diff
@@ -154,11 +186,144 @@ async function createOutgoingPayment(
       }
     } else if (isOutgoingPaymentError(err)) {
       return err
+    } else if (err instanceof knex.KnexTimeoutError) {
+      deps.logger.error({ grant: options.grant.grant }, 'grant locked')
     }
     throw err
   }
 }
 
+function validateAccessLimits(
+  payment: OutgoingPayment,
+  limits: AccessLimits
+): PaymentLimits | undefined {
+  if (
+    (!limits.receiver || payment.receiver === limits?.receiver) &&
+    validateAmountAssets(payment, limits)
+  ) {
+    let paymentInterval: Interval | undefined
+    if (limits.interval)
+      paymentInterval = getInterval(limits.interval, payment.createdAt)
+    if (!limits.interval || !!paymentInterval) {
+      return { ...limits, paymentInterval }
+    }
+  }
+}
+
+function validatePaymentInterval({
+  limits,
+  payment
+}: {
+  limits: PaymentLimits
+  payment: OutgoingPayment
+}): boolean {
+  return (
+    !limits.paymentInterval ||
+    (limits.paymentInterval.start.toMillis() <= payment.createdAt.getTime() &&
+      payment.createdAt.getTime() < limits.paymentInterval.end.toMillis())
+  )
+}
+
+function validateAmountAssets(
+  payment: OutgoingPayment,
+  limits: AccessLimits
+): boolean {
+  if (
+    limits.sendAmount &&
+    (limits.sendAmount.assetCode !== payment.asset.code ||
+      limits.sendAmount.assetScale !== payment.asset.scale)
+  ) {
+    return false
+  }
+  return (
+    !limits.receiveAmount ||
+    (limits.receiveAmount.assetCode === payment.receiveAmount.assetCode &&
+      limits.receiveAmount.assetScale === payment.receiveAmount.assetScale)
+  )
+}
+
+interface PaymentLimits extends AccessLimits {
+  paymentInterval?: Interval
+}
+
+// "payment" is locked by the "deps.knex" transaction.
+async function validateGrant(
+  deps: ServiceDependencies,
+  payment: OutgoingPayment,
+  grant: Grant
+): Promise<boolean> {
+  const grantAccess = grant.access[0]
+  if (!grantAccess.limits) {
+    return true
+  }
+  const paymentLimits = validateAccessLimits(payment, grantAccess.limits)
+  if (!paymentLimits) {
+    return false
+  }
+  if (!paymentLimits.sendAmount && !paymentLimits.receiveAmount) return true
+  if (
+    (paymentLimits.sendAmount &&
+      paymentLimits.sendAmount.value < payment.sendAmount.value) ||
+    (paymentLimits.receiveAmount &&
+      paymentLimits.receiveAmount.value < payment.receiveAmount.value)
+  ) {
+    // Payment amount single-handedly exceeds amount limit
+    return false
+  }
+
+  const grantPayments = await OutgoingPayment.query(deps.knex)
```

Comment:
:thinking: Technically, we don't need to lock the grant until just before we perform this query.

Reply:
That is correct but in the realm of what was said this morning, I'd rather keep it as is. I don't think it makes a big difference.

Reply:
:+1: I opened an issue for it:
- https://github.com/interledger/rafiki/issues/485